src/Streaming/Internal.hs

```diff
@@ -727,13 +779,40 @@ replicates n f = splitsAt n (repeats f) >> return ()
 cycles :: (Monad m, Functor f) =>  Stream f m () -> Stream f m r
 cycles = forever
 
+-- | A version of 'hoist' that works properly even when its
+-- argument is not a monad morphism.
+--
+-- > hoistUnexposed = hoist . unexposed
+hoistUnexposed :: (Monad m, Functor f) => (forall a. m a -> n a) -> Stream f m r -> Stream f n r
+hoistUnexposed trans = Effect . loop where
+  loop stream = trans $ do
+    rs <- inspect stream
+    case rs of
+      Left r -> return (Return r)
+      Right f -> return (Step (fmap (Effect . loop) f))
+{-# INLINABLE hoistUnexposed #-}
```

Comment:
This can surely be made a tad more efficient by using the `inspectC` function I designed to implement `zipsWith`. If we merge that first, we can use the function here.